omniledger/service/service.go

```diff
@@ -475,76 +473,71 @@ func (s *Service) verifySkipBlock(newID []byte, newSB *skipchain.SkipBlock) bool
+		return false
 	}
+	ctx := body.Transactions
 	cdb := s.getCollection(newSB.Hash)
-	mtr, ots := s.createOmniledgerTransactions(cdb.coll, ctx)
-	if bytes.Compare(mtr, header.CollectionRoot) != 0 {
-		log.Lvl2(s.ServerIdentity(), "Collection root doesn't verify")
+	mtr, _, scs, err := s.createStateChanges(cdb.coll, ctx)
+	if err != nil {
+		log.Error("Couldn't create state changes:", err)
 		return false
 	}
-	hash := sha256.New()
-	for _, tx := range ots {
-		txBuf, err := protobuf.Encode(&tx)
-		if err != nil {
-			log.Lvl2(s.ServerIdentity(), "Couldn't marshal transaction")
-		}
-		hash.Write(txBuf)
+	if bytes.Compare(header.CollectionRoot, mtr) != 0 {
+		log.Lvl2(s.ServerIdentity(), "Collection root doesn't verify")
+		return false
 	}
-	if bytes.Compare(hash.Sum(nil), header.TransactionHash) != 0 {
-		log.Lvl2(s.ServerIdentity(), "Transaction hash doesn't verify")
+	if bytes.Compare(header.StateChangesHash, scs.Hash()) != 0 {
+		log.Lvl2(s.ServerIdentity(), "State Changes hash doesn't verify")
 		return false
 	}
 	return true
 }
 
-// createOmniledgerTransactions goes through all ClientTransactions and creates
-// the appropriate StateChanges and sets the valid flag.
-func (s *Service) createOmniledgerTransactions(coll collection.Collection, txs []ClientTransaction) ([]byte, []OmniledgerTransaction) {
+// createStateChanges goes through all ClientTransactions and creates
+// the appropriate StateChanges. If any of the transactions are invalid,
+// it returns an error.
+func (s *Service) createStateChanges(coll collection.Collection, cts ClientTransactions) (merkleRoot []byte, ctsOK ClientTransactions, states StateChanges, err error) {
 	cdbTemp := coll.Clone()
-	var otx []OmniledgerTransaction
 clientTransactions:
-	for _, ctx := range txs {
+	for _, ct := range cts {
 		cdbI := cdbTemp.Clone()
-		ot := OmniledgerTransaction{ClientTransaction: ctx, Valid: true}
-		for _, instr := range ctx.Instructions {
-			kind, state, err := instr.GetKindState(cdbI)
+		for _, instr := range ct.Instructions {
+			kind, _, err := instr.GetContractState(cdbI)
 			if err != nil {
 				log.Lvl1("Couldn't get kind of instruction")
 				continue clientTransactions
 			}
 
-			f, exists := s.classes[kind]
+			f, exists := s.contracts[kind]
 			// If the leader does not have a verifier for this kind, it drops the
 			// transaction.
 			if !exists {
 				log.Lvl1("Leader is dropping instruction of unknown kind:", kind)
 				continue clientTransactions
 			}
-			// Now we call the class function with the data of the key:
-			scs, err := f(cdbI, instr, kind, state)
+			// Now we call the contract function with the data of the key:
+			log.Lvlf3("%s: Calling contract %s", s.ServerIdentity(), kind)
+			scs, _, err := f(cdbI, instr, nil)
 			if err != nil {
 				log.Lvl1("Call to class returned error:", err)
-				ot.Valid = false
-				ot.StateChanges = nil
 				cdbI = cdbTemp
-				break
+				continue clientTransactions
```

Comment:
so we're not storing bad transactions anymore?

Reply:
No, because we want to be able to avoid spamming by a malicious client, so we need to make sure that he has at least enough coins. Even though he could still send a lot of transactions, but at least they will not be included in the block.